core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -0,0 +1,270 @@
+                filterSplitResult.getPushableConditions(),
+                // TODO we could send only subset of assignments here, those which are needed to resolve filterSplitResult.getPushableConditions
+                leftAssignments,
+                rightAssignments);
+
+        if (joinApplicationResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        TableHandle handle = joinApplicationResult.get().getTableHandle();
+
+        Map<ColumnHandle, ColumnHandle> columnHandlesMapping = joinApplicationResult.get().getColumnHandles();
+
+        TupleDomain<ColumnHandle> newEnforcedConstraints = TupleDomain.fromColumnDomains(
+                transformToNewAssignments(left.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
+                        leftDomains -> transformToNewAssignments(right.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
+                                rightDomains -> Optional.of(ImmutableList.<TupleDomain.ColumnDomain<ColumnHandle>>builder().addAll(leftDomains).addAll(rightDomains).build()))));
+
+        ImmutableMap.Builder<Symbol, ColumnHandle> newAssignments = ImmutableMap.builder();
+        newAssignments.putAll(left.getAssignments().entrySet().stream().collect(toImmutableMap(
+                Map.Entry::getKey,
+                entry -> columnHandlesMapping.get(entry.getValue()))));
+        newAssignments.putAll(right.getAssignments().entrySet().stream().collect(toImmutableMap(
+                Map.Entry::getKey,
+                entry -> columnHandlesMapping.get(entry.getValue()))));
+
+        return Result.ofPlanNode(new TableScanNode(left.getId(), handle, joinNode.getOutputSymbols(), newAssignments.build(), newEnforcedConstraints, false));
+    }
+
+    private TupleDomain<ColumnHandle> transformToNewAssignments(TupleDomain<ColumnHandle> tupleDomain, Map<ColumnHandle, ColumnHandle> newAssignments)
+    {
+        return tupleDomain.transform(handle -> {
+            ColumnHandle newHandle = newAssignments.get(handle);
+            if (newHandle == null) {
+                throw new IllegalArgumentException("Mapping not found for handle " + handle);
+            }
+            return newHandle;
+        });
+    }
+
+    public Expression getEffectiveFilter(JoinNode node)
+    {
+        Expression effectiveFilter = and(node.getCriteria().stream().map(JoinNode.EquiJoinClause::toExpression).collect(toImmutableList()));
+        if (node.getFilter().isPresent()) {
+            effectiveFilter = and(effectiveFilter, node.getFilter().get());
+        }
+        return effectiveFilter;
+    }
+
+    private FilterSplitResult splitFilter(Expression filter, List<Symbol> leftSymbols, List<Symbol> rightSymbols, Context context)
+    {
+        Set<String> leftSymbolNames = leftSymbols.stream().map(Symbol::getName).collect(toImmutableSet());
+        Set<String> rightSymbolNames = rightSymbols.stream().map(Symbol::getName).collect(toImmutableSet());
```

Comment:
You can make `Set<Symbol>`, this will make code shorter

Reply:
Changed - I hope you meant that.